lib_eio/unix/pty.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2004 Anil Madhavapeddy <anil@recoil.org>
+ * Copyright (c) 2020–2021 Craig Ferguson <me@craigfe.io>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+
+#include <stdio.h>
+#include <errno.h>
+#include <paths.h>
+#include <fcntl.h>
+#include <string.h>
+#include <termios.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <signal.h>
+#include <limits.h>
+
+#include <pty.h>
+#include <utmp.h>
+
+#include <caml/mlvalues.h>
+#include <caml/alloc.h>
+#include <caml/memory.h>
+#include <caml/fail.h>
+#include <caml/callback.h>
+#include <caml/signals.h>
+#include <caml/unixsupport.h>
+
+value eio_unix_open_pty(value v_unit)
+{
+    CAMLparam1 (v_unit);
+    char namebuf[4096]; /* Not PATH_MAX due to portability issues */
+    int i, masterfd, slavefd;
+    CAMLlocal1(v_ret);
+
+    i = openpty(&masterfd, &slavefd, namebuf, NULL, NULL);
```

Comment:
The man-page notes that:
> Nobody knows how much space should be reserved for name.  So, calling openpty() or forkpty() with non-NULL name may  not be secure.

However, `ptsname_r` is Linux-only.

In a multi-domain program, all FDs must be created close-on-exec atomically, or they can leak. https://www.austingroupbugs.net/view.php?id=411 says:
> The function posix_openpt( ) with the O_CLOEXEC flag is necessary to
avoid a data race in multi-threaded applications. Without it, a file
descriptor is leaked into a child process created by one thread in the
window between another thread creating a file descriptor with
posix_openpt( ) and then using fcntl( ) to set the FD_CLOEXEC flag.

However, it's not clear how widely-supported that is.

The man-page for `posix_openpt` notes that you can just open `/dev/ptmx` directly, which would avoid these problems, but require some other manual setup.

Reply:
It seems to come down to using either BSD-style pty handling (openpty) or Unix98-style handling (posix_openpt).  The latter seems very much preferred, and a quick smoke test on macOS, Linux and OpenBSD shows posix_openpt is present there.  I'll have a go at switching this PR to using /dev/ptmx instead -- there's code in portable OpenSSH to use as a guide.